compiler/plugins/target/AMD-AIE/aie/AMDAIEObjectFifoStatefulTransform.cpp

```diff
@@ -65,1279 +52,918 @@ class LockAnalysis {
+
+  auto fifo = llvm::cast<AIEObjectFifoType>(op.getElemType());
+  auto elemType = llvm::cast<MemRefType>(fifo.getElementType());
+  int64_t len = elemType.getNumElements();
+  createMemDMA<MemOp, BufferOp>(device, builder, op, target, channelDir,
+                                channelIndex, dims, numBlocks,
+                                /*acqNum*/ 1, /*relNum*/ 1, len, /*offset*/ 0,
+                                buffersPerFifo, locksPerFifo);
+}
 
-  /// Function used to create a MemOp region with a DMA channel.
-  /// It uses creatBdBlock(), see there for lockMode input.
-  void createAMDAIETileDMA(DeviceOp &device, OpBuilder &builder,
-                           ObjectFifoCreateOp op, DMAChannelDir channelDir,
-                           int channelIndex, int lockMode,
-                           BDDimLayoutArrayAttr dims) {
-    size_t numBlocks = op.size();
-    if (numBlocks == 0) return;
-
-    int acqNum = 1;
-    int relNum = 1;
-
-    auto fifo = llvm::cast<AIEObjectFifoType>(op.getElemType());
-    auto elemType = llvm::cast<MemRefType>(fifo.getElementType());
-    int len = elemType.getNumElements();
-
-    // search for the buffers/locks (based on if this objFifo has a link)
-    ObjectFifoCreateOp target = op;
-    if (std::optional<ObjectFifoLinkOp> linkOp = getOptionalLinkOp(op);
-        linkOp.has_value())
-      if (objFifoLinks.find(linkOp.value()) != objFifoLinks.end())
-        target = objFifoLinks[linkOp.value()];
-
-    // search for MemOp
-    Operation *producerMem = nullptr;
-    for (auto memOp : device.getOps<MemOp>()) {
-      if (memOp.getTile() == op.getProducerTile()) {
-        producerMem = memOp.getOperation();
-        break;
+void createMemTileDMA(
+    DeviceOp &device, OpBuilder &builder, ObjectFifoCreateOp op,
+    DMAChannelDir channelDir, int channelIndex, BDDimLayoutArrayAttr dims,
+    DenseMap<ObjectFifoLinkOp, ObjectFifoCreateOp> &objFifoLinks,
+    DenseMap<ObjectFifoCreateOp, std::vector<BufferOp>> &buffersPerFifo,
+    DenseMap<ObjectFifoCreateOp, std::vector<LockOp>> &locksPerFifo) {
+  size_t numBlocks = op.size();
+  if (numBlocks == 0) return;
+
+  auto fifo = llvm::cast<AIEObjectFifoType>(op.getElemType());
+  auto elemType = llvm::cast<MemRefType>(fifo.getElementType());
+  int64_t lenOut = elemType.getNumElements();
+  size_t acqNum = 1;
+  size_t relNum = 1;
+  int64_t extraOffset = 0;
+  ObjectFifoCreateOp target = op;
+
+  auto getExtraOffset = [&acqNum, &relNum, &extraOffset, &target, &op](
+                            ObjectFifoLinkOp linkOp,
+                            const std::vector<ObjectFifoCreateOp> &fifos,
+                            size_t size) {
+    if (target == op) {
+      acqNum = size;
+      relNum = size;
+    } else
+      for (auto fifoIn : fifos) {
+        auto fifoType = llvm::cast<AIEObjectFifoType>(fifoIn.getElemType());
+        auto fifoElemType = llvm::cast<MemRefType>(fifoType.getElementType());
+        if (fifoIn.name() == op.name()) break;
```

Comment:
What is this for?

Reply:
![image](https://github.com/nod-ai/iree-amd-aie/assets/5657668/0d45186e-1b47-496b-b9b9-668f0f8cebfd)